### PR TITLE
feat: log timing and attempt details

### PR DESCRIPTION
**Pull request by maxmoehl**
This commit adds additional fields to the access and error log to aid with troubleshooting performance issues and better understanding of the time spent in the system.

Access log fields:
* `failed_attempts`: number of attempts that failed
* `failed_attempts_time`: time spent on attempts that failed
* `dns_time`: time spent looking up DNS records
* `dial_time`: time spent dialing the network connection
* `tls_time`: time spent performing the TLS handshake
* `backend_time`: time spent waiting for the backend to respond

Error log fields:
* `backend-time`: time spent waiting for the backend to respond

Resolves: cloudfoundry/routing-release#304

* [X] I have viewed signed and have submitted the Contributor License Agreement
* [X] I have made this pull request to the `main` branch
* [X] I have run all the unit tests using `scripts/run-unit-tests-in-docker` from [routing-release](https://github.com/cloudfoundry/routing-release).
  ```
  Summarizing 1 Failure:
  [FAIL] Router Integration prometheus metrics [It] starts a prometheus https server
  /routing-release/src/code.cloudfoundry.org/gorouter/integration/main_test.go:732
  ```
  It seems like the prometheus setup is not working in the docker container, but this change is unrelated to prometheus.
* [ ] (Optional) I have run Routing Acceptance Tests and Routing Smoke Tests on bosh lite
* [ ] (Optional) I have run CF Acceptance Tests on bosh lite

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -148,6 +148,7 @@ type LoggingConfig struct {
 	DisableLogForwardedFor bool         `yaml:"disable_log_forwarded_for"`
 	DisableLogSourceIP     bool         `yaml:"disable_log_source_ip"`
 	RedactQueryParams      string       `yaml:"redact_query_params"`
+	AttemptsDetails        bool         `yaml:"attempts_details"`
```
Comment (domdom82) [inline config/config.go:151]: since this is a flag, I'd put this in verbal form like the other flags.
e.g.
```
LogAttemptsDetails
EnableAttemptsDetails
UseAttemptsDetails
LogDetailedTimings (see access_log.go where you did that, too)
```
etc.

File: accesslog/schema/access_log_record.go [hunk 2]
```diff
@@ -97,23 +97,27 @@ func (b *recordBuffer) WriteDashOrStringValue(s string) {
...
+	AppRequestStartedAt        time.Time
+	AppRequestFinishedAt       time.Time
+	BodyBytesSent              int
+	RequestBytesReceived       int
+	ExtraHeadersToLog          []string
+	DisableXFFLogging          bool
+	DisableSourceIPLogging     bool
+	RedactQueryParams          string
+	RouterError                string
+	LogAttemptsDetails         bool
```
Comment (domdom82) [inline accesslog/schema/access_log_record.go:116]: didn't we say we wanted one `AttemptsDetails` field here that could later be extended?
we are now adding 5 single fields plus one additional flag.

File: handlers/access_log.go [hunk 3]
```diff
@@ -65,6 +69,11 @@ func (a *accessLog) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http
 	alr.BodyBytesSent = proxyWriter.Size()
 	alr.StatusCode = proxyWriter.Status()
 	alr.RouterError = proxyWriter.Header().Get(router_http.CfRouterError)
+	alr.DNSTime = reqInfo.DNSTime
```
Comment (domdom82) [inline handlers/access_log.go:72]: as commented above, wouldn't it be easier to maintain if we just added one `AttemptsDetails` struct to the reqInfo?

File: accesslog/schema/access_log_record_test.go [hunk 4]
```diff
@@ -418,4 +418,78 @@ var _ = Describe("AccessLogRecord", func() {
...
+
+			Expect(r).To(ContainSubstring("failed_attempts:4"))
+			Expect(r).To(ContainSubstring("failed_attempts_time:1.0"))
+			Expect(r).To(ContainSubstring("dns_time:1.2"))
+			Expect(r).To(ContainSubstring("dial_time:1.3"))
+			Expect(r).To(ContainSubstring("tls_time:1.5"))
+			Expect(r).To(ContainSubstring("backend_time:1.0"))
+		})
+
+		It("adds all appropriate empty values if fields are unset", func() {
```
[the remark this answers is not in the document]
Comment (domdom82) [reply accesslog/schema/access_log_record_test.go:471]: There should also be a test for negative `backend_time` to test the `BackendTime` function of reqInfo

File: proxy/round_tripper/proxy_round_tripper.go [hunk 5]
```diff
@@ -246,6 +259,17 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
...
+	reqInfo.FailedAttempts = attempt - 1
+
+	if reqInfo.FailedAttempts > 0 && lastFailedAttempt.IsZero() {
+		panic("forgot to record end of failed attempts")
+	}
+
+	if !lastFailedAttempt.IsZero() {
+		reqInfo.FailedAttemptsTime = lastFailedAttempt.Sub(beginAttempts).Seconds()
+	}
+
```
Comment (domdom82) [inline proxy/round_tripper/proxy_round_tripper.go:272]: why don't you replace this complexity with a simple
```
reqInfo.FailedAttempts++
reqInfo.FailedAttemptsTime = time.Now().Sub(beginAttempts).Seconds()
```
in the two cases?

Comment (maxmoehl) [reply proxy/round_tripper/proxy_round_tripper.go:272]: We considered this but it feels very brittle, we must increment the counter at different places and any future change must consider this. This way we rely on the attempts counter which is known to be correct and only modified in a single place. I'm also very unhappy with the way we set `lastFailedAttempt` because it suffers from the same issue (hence the panic to hopefully prevent such conditions by failing loud and clear during testing), not sure where I would put the second LOC you suggested.

File: accesslog/schema/access_log_record.go (hunk 2 above)
Comment (maxmoehl) [reply accesslog/schema/access_log_record.go:116]: Discussed offline: we will leave it like this for now but re-consider this overall structure later.

File: proxy/round_tripper/proxy_round_tripper.go (hunk 5 above)
Comment (maxmoehl) [reply proxy/round_tripper/proxy_round_tripper.go:272]: After further discussion we agreed to move add the two lines to each error case and remove the checks after the loop.

File: accesslog/schema/access_log_record_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxmoehl) [reply accesslog/schema/access_log_record_test.go:471]: The test for `BackendTime` is done.

File: config/config.go [hunk 6]
```diff
@@ -148,6 +148,7 @@ type LoggingConfig struct {
 	DisableLogForwardedFor bool         `yaml:"disable_log_forwarded_for"`
 	DisableLogSourceIP     bool         `yaml:"disable_log_source_ip"`
 	RedactQueryParams      string       `yaml:"redact_query_params"`
+	EnableAttemptsDetails  bool         `yaml:"attempts_details"`
```
[the remark this answers is not in the document]
Comment (domdom82) [reply config/config.go:151]: probably also affects config tests

**Comment by maxmoehl** [comment]
Before merging I will squash the commits and adapt the commit message accordingly.

File: proxy/round_tripper/proxy_round_tripper.go [hunk 7]
```diff
@@ -255,15 +286,26 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
...
-	if finalErr != nil {
-		rt.errorHandler.HandleError(reqInfo.ProxyResponseWriter, finalErr)
-		return nil, finalErr
+	if err != nil {
+		rt.errorHandler.HandleError(reqInfo.ProxyResponseWriter, err)
+		return nil, err
 	}
+
+	// Record the times from the last attempt, but only if it succeeded.
+	reqInfo.DnsStartedAt = time.Unix(0, trace.tDNSStart.Load())
```
Comment (domdom82) [inline proxy/round_tripper/proxy_round_tripper.go:302]: I'd prefer having this in the requestTracer type.
e.g.
```
reqInfo.DnsStartedAt = trace.DnsStartedAt()
```
and remove the `trace.DnsTime()` etc. functions as we no longer need them since we only record timestamps now.

File: accesslog/schema/access_log_record.go [hunk 8]
```diff
@@ -101,40 +101,98 @@ type AccessLogRecord struct {
...
+		return -1
+	}
+	return r.LastFailedAttemptFinishedAt.Sub(r.AppRequestStartedAt).Seconds()
+}
+
+// successfulAttemptTime returns -1 if there was an error, so no attempt was
+// successful. If there was a successful attempt the returned time indicates
+// how long it took.
+func (r *AccessLogRecord) successfulAttemptTime() float64 {
+	if r.AppRequestFinishedAt.UnixNano() == r.LastFailedAttemptFinishedAt.UnixNano() {
```
Comment (domdom82) [inline accesslog/schema/access_log_record.go:185]:
```suggestion
	if r.AppRequestFinishedAt.Equal(r.LastFailedAttemptFinishedAt) {
```
?

File: proxy/round_tripper/proxy_round_tripper.go (hunk 7 above)
Comment (maxmoehl) [reply proxy/round_tripper/proxy_round_tripper.go:302]: We are still using those for the error logs, so I will keep them.

**Review by domdom82 — APPROVED** [review]
I like it!
